### PR TITLE
Multilayer amr

### PR DESCRIPTION
**Pull request by aks2203**

### CODE REVIEW

**Comment by Dirivian** [comment]
While running a few examples (Plane wave and radial wave), I get an error I previously encountered in multilayer-examples  in the apps repo.
`Error: Function ‘rho’ at (1) has no IMPLICIT type
` `
/home/jithin/clawpack_src/clawpack-v5.4.1/clawutil/src/Makefile.common:127: recipe for target 'qinit_module.mod' failed
make: *** [qinit_module.mod] Error 1
`
It is unable to find the  'rho' in multilayer_module. Unsure if this is unique to me, right now.

**Comment by mandli** [comment]
Hi Nithin,

Did you just upgrade to a new version?  The parameter `rho` was moved to the geoclaw module recently and left overs from previous compilations may be causing that.

Kyle

**Comment by Dirivian** [comment]
Yes, I think I installed my current version of clawpack mid-July this year.

**Comment by mandli** [comment]
I might suggest then to remove all *.mod and *.o files from the source directories. Check to make sure that there are no local files with those suffixes. See if that fixes the problem.

Kyle

**Comment by Dirivian** [comment]
I tried that with no effect initially. Then, I replaced the geoclaw in my installation with this one and this error didn't appear. However, the following one did appear.

`Error: Symbol ‘timefilvaltot’ referenced at (1) not found in module ‘amr_module`

I cloned the latest version of amrclaw (without the .o and .mod files) and put it in my installation path. But, it doesn't change anything .

Jithin

**Comment by rjleveque** [comment]
I run into the same problem as @Dirivian.  I think the problem is that this branch `aks2203:conflict-resolution` is out of date with `master`.  In particular, `timeFilvalTot` was removed from `amr2.f90` in commit 7a39383b6 and from `amr_module.f90` in clawpack/amrclaw@bd0079c685 but is still in `amr2.f90` on this branch.

@aks2203: Maybe we can sit down together next week and sort this out.

**Comment by mandli** [comment]
If the error involves `rho` though I don't think that's an issue with `timeFilvalTot` though is it?

**Comment by Dirivian** [comment]
@mandli The `rho` error was something I managed to remove/circumvent with your suggestions. The `timeFilvalTot` error is something that came after that.

**Comment by mandli** [comment]
@Dirivian Ah, ok.  Then I assume more generally then we just need to figure out the difference in `update.f90`.

**Comment by aks2203** [comment]
This last failing check is not a nosetests problem. The "60% mismatch" has been fixed. It was a bug in my code leftover from implicit declaration of integers in the old version update.f.

**Comment by mandli** [comment]
What did you need to change?  Does it require another repository?

**Comment by aks2203** [comment]
No it does not. It was two variable in update2.f90 which were declared the wrong way. I have fixed them and pushed it.

File: examples/multi-layer/bowl-radial/README.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
+
+.. _geoclaw_examples_tsunami_bowl-radial:
+
+Tsunami in a parabolic bowl shaped ocean
+========================================
+
+Waves in a parabolic bowl generated by a Gaussian hump of water at the
+origin.  The solution should remain circularly symmetric.
+
```
Comment (mandli) [inline examples/multi-layer/bowl-radial/README.rst]: Add that this is a multi-layer version

File: examples/multi-layer/chile2010_ml/Makefile [hunk 2]
```diff
@@ -0,0 +1,111 @@
+# Makefile for Clawpack code in this directory.
```
Comment (mandli) [inline examples/multi-layer/chile2010_ml/Makefile]: Probably do not need `_ml` directory name given it's in the multi-layer directory.

File: examples/multi-layer/chile2010_ml/README.rst [hunk 3]
```diff
@@ -0,0 +1,46 @@
+
+.. _geoclaw_examples_tsunami_chile2010:
+
+Tsunami arising offshore Maule, Chile, Feb. 27, 2010 
+=====================================================
+
```
Comment (mandli) [inline examples/multi-layer/chile2010_ml/README.rst]: Again add multi-layer statement

File: examples/multi-layer/chile2010_ml/genline.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+fh = open('line.xyz', 'w')
```
Comment (mandli) [inline examples/multi-layer/chile2010_ml/genline.py]: What is this for?

File: examples/multi-layer/chile2010_ml/makehump.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+"""
```
Comment (mandli) [inline examples/multi-layer/chile2010_ml/makehump.py]: Is this the equivalent to `makeqinit.py`?  Can this be added to `setrun.py`.

File: examples/multi-layer/chile2010_ml/maketopo.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
+"""
```
Comment (mandli) [inline examples/multi-layer/chile2010_ml/maketopo.py]: If possible also add this to `setrun.py`.

File: examples/multi-layer/chile2010_ml/plotfg.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (mandli) [inline examples/multi-layer/chile2010_ml/plotfg.py]: Was this in the original chile example?

File: examples/multi-layer/plane_wave/Makefile [hunk 8]
```diff
@@ -39,9 +39,9 @@ else ifeq ($(UNAME), Linux)
 		# for other installation types
 		FFLAGS += -I$(MKLROOT)/include
 		ifndef LFLAGS                                                                                                                                    
-		    LFLAGS =  -Wl,--start-group $(MKLROOT)/lib/intel64/libmkl_intel_lp64.a $(MKLROOT)/lib/intel64/libmkl_core.a $(MKLROOT)/lib/intel64/libmkl_sequential.a -Wl,--end-group -lpthread -lm $(FFLAGS)
+			LFLAGS =  -Wl,--start-group $(MKLROOT)/lib/intel64/libmkl_intel_lp64.a $(MKLROOT)/lib/intel64/libmkl_core.a $(MKLROOT)/lib/intel64/libmkl_sequential.a -Wl,--end-group -lpthread -lm $(FFLAGS)
```
Comment (mandli) [inline examples/multi-layer/plane_wave/Makefile]: Looks like an extraneous tab is in here.

File: examples/multi-layer/square_multi/Makefile [hunk 9]
```diff
@@ -0,0 +1,124 @@
+
```
Comment (mandli) [inline examples/multi-layer/square_multi/Makefile]: Containing directory is misspelled.  I might suggest just shortening it to `square` or `square_basin`.

File: src/2d/shallow/filpatch2.f90 [hunk 10]
```diff
@@ -0,0 +1,496 @@
+! :::::::::::::::::::::::::::: FILPATCH :::::::::::::::::::::::::;
```
Comment (mandli) [inline src/2d/shallow/filpatch2.f90]: Is there a `filpatch.f90`?

File: examples/multi-layer/chile2010_ml/plotfg.py (hunk 7 above)
Comment (aks2203) [reply examples/multi-layer/chile2010_ml/plotfg.py]: Yes.

**Comment by mandli** [comment]
I think the only thing left to do is either change the current multi-layer test to include AMR or create a new test based on one of the new examples preferring the latter.

**Comment by mandli** [comment]
I just ran this branch and found that OpenMP is broken:
```
/Users/mandli/src/clawpack/geoclaw/src/2d/shallow/update.f90:46:64:

 !$OMP PARALLEL DO PRIVATE(ng,mptr,loc,loccaux,nx,ny,mitot,mjtot,
                                                                1
Error: Syntax error in OpenMP variable list at (1)
```
I have not had a chance to really figure out what's wrong.  I imagine comparing to the original `update.f` should show the problem though.

**Comment by mandli** [comment]
Never mind, I think there may be multiple problems with this OpenMP statement.  There are also a bunch of variables in the block that are not defined.

**Comment by aks2203** [comment]
I pushed prematurely, I'm still working on new tests for multilayer_amr.

File: examples/multi-layer/chile_2010/plot_dart.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+pylab.hold(True)
+
+# origin of plot in pixels on image:
+t0 = 195.
+y0 = 221.
+tscale = 120.0
+yscale = -446.66666666666669
+plotdata = ClawPlotData()
+plotdata.outdir = "_output"
+gaugedata = plotdata.getgauge(32412)
```
[the remark this answers is not in the document]
Comment (aks2203) [reply examples/multi-layer/chile_2010/plot_dart.py]: It doesn't work for me. I must have copied it from another directory. Should it just be removed? This example doesn't run anyway, perhaps the whole example shouldn't be merged in if it fails almost immediately.

[the remark this answers is not in the document]
Comment (mandli) [reply examples/multi-layer/chile_2010/plot_dart.py]: I would say remove it.  Remove the example may not be a bad idea, at least until we can get it working.  Will avoid having to explain to people who try to use it.

**Comment by mandli** [comment]
@aks2203 I just fixed a minor conflict with a new PR I just merged, can you check to make sure I did that correctly.

File: src/2d/shallow/multilayer/flag2refine2.f90 [hunk 12]
```diff
@@ -134,54 +174,44 @@ subroutine flag2refine2(mx,my,mbc,mbuff,meqn,maux,xlower,ylower,dx,dy,t,level, &
...
-                    if (h(k) > dry_tolerance(k)) then
-                        eta = h(k) + eta_below
-                    
                         ! Check wave criteria
-                        if (abs(eta - eta_init(k)) > wave_tolerance(k)) then
+                        if (abs(eta - eta_init(layer)) > wave_tolerance(layer)) then
                             ! Check to see if we are near shore
-                            if (total_depth < deep_depth) then
+                            if (q(3*layer-2,i,j) / rho(layer) < deep_depth) then
+!                                 print *, "1"
```
Comment (mandli) [inline src/2d/shallow/multilayer/flag2refine2.f90]: Can you delete these debugging print statements?

File: src/python/geoclaw/data.py [hunk 13]
```diff
@@ -321,14 +321,15 @@ def write(self,data_source='setrun.py'):
         # Perturbation requested
         if self.qinit_type == 0:
             pass
-        elif 0 < self.qinit_type < 5:
+        # elif 0 < self.qinit_type < 5:
```
Comment (mandli) [inline src/python/geoclaw/data.py]: Might as well remove these.

**Comment by mandli** [comment]
The one thing I am wondering now is if we should really try to merge the changes to `filval` and others into the single-layer code.  @aks2203 did you have any luck with the timing tests for that?

**Comment by mandli** [comment]
The remaining issue regarding greater integration of this work has been raised an issue in #284.

**Pull request merged**